### PR TITLE
Service and controller tests added for ShoppingCart model

### PR DESCRIPTION
**Pull request by jszymanski97**

### CODE REVIEW

File: src/test/java/mate/project/service/ShoppingCartServiceTest.java
Comment (George-Khomiak) [inline src/test/java/mate/project/service/ShoppingCartServiceTest.java:1]: Naming convention is off, for example many tests have **ShoppingCartResponseDto**() at the end, but this is just the return type, add smth like returnsShoppingCartResponseDto.
Or the method like **deleteCartItemById_validCartItemId_void**() this "void" at the end of the name has no informative value, it should explain what should happen as the result of test case execution.

File: src/test/java/mate/project/service/ShoppingCartServiceTest.java [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+    }
+
+    @Test
+    @DisplayName("Finding all shopping carts for authenticated user")
+    void findAll_validUser_ShoppingCartResponseDto() {
+        // Given
+        when(userRepository.findByEmail("testUser")).thenReturn(Optional.of(user));
+        when(shoppingCartRepository.findAllByUserId(user.getId()))
+                .thenReturn(List.of(shoppingCart));
+        when(shoppingCartRepository.findAllByUserId(user.getId()))
```
Comment (George-Khomiak) [inline src/test/java/mate/project/service/ShoppingCartServiceTest.java]: this line is duplicated

File: src/test/java/mate/project/service/ShoppingCartServiceTest.java [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+
+        // When & Then
+        assertThrows(EntityNotFoundException.class, () -> shoppingCartService
+                .deleteCartItemById(999L));
+        verify(shoppingCartRepository, times(0)).save(shoppingCart);
+    }
+
+    @Test
+    @DisplayName("Finding all shopping carts for authenticated user")
+    void findAll_validUser_ShoppingCartResponseDto() {
```
Comment (George-Khomiak) [inline src/test/java/mate/project/service/ShoppingCartServiceTest.java]: Let's add one test to check invalid user

File: src/test/java/mate/project/service/ShoppingCartServiceTest.java [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+        shoppingCartService.deleteCartItemById(cartItem.getId());
+
+        // Then
+        assertEquals(0, shoppingCart.getCartItems().size());
+        verify(shoppingCartRepository, times(1)).save(shoppingCart);
+    }
+
+    @Test
+    @DisplayName("Deleting a non-existing shopping cart item")
+    void deleteCartItemById_nonExistingCartItem_throwsEntityNotFoundException() {
```
Comment (George-Khomiak) [inline src/test/java/mate/project/service/ShoppingCartServiceTest.java:178]: This test name is very clear

File: src/test/java/mate/project/service/ShoppingCartServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+        Authentication authentication = mock(Authentication.class);
+        SecurityContext securityContext = mock(SecurityContext.class);
+        when(securityContext.getAuthentication()).thenReturn(authentication);
+        when(authentication.getName()).thenReturn("testUser");
+        SecurityContextHolder.setContext(securityContext);
+    }
+
+    @Test
+    @DisplayName("Adding a new shopping cart item")
+    void addCartItem_validCartItemRequestDto_ShoppingCartResponseDto() {
```
Comment (George-Khomiak) [inline src/test/java/mate/project/service/ShoppingCartServiceTest.java]: Let's check what happens when some null values are provided, add one more edge case

File: src/test/java/mate/project/controller/ShoppingCartControllerTest.java [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+
+    @BeforeEach
+    void setup() {
+        mockMvc = MockMvcBuilders
+                .webAppContextSetup(webApplicationContext)
+                .apply(springSecurity())
+                .build();
+    }
+
+    @WithMockUser(username = "user1@example.com", roles = {"USER"})
```
Comment (George-Khomiak) [inline src/test/java/mate/project/controller/ShoppingCartControllerTest.java]: extract user names into constants

File: src/test/java/mate/project/controller/ShoppingCartControllerTest.java [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+        sortedCartItems.sort(Comparator.comparing(CartItemResponseDto::getBookId));
+        // Converted set to list to allow sorting and checking the books in cart
+
+        CartItemResponseDto firstItem = sortedCartItems.get(0);
+        Assertions.assertEquals("The Hobbit", firstItem.getBookTitle());
+        Assertions.assertEquals(2, firstItem.getQuantity());
+    }
+
+    @WithMockUser(username = "user1@example.com", roles = {"USER"})
+    @DisplayName("Fetch shopping cart for user1")
```
Comment (George-Khomiak) [inline src/test/java/mate/project/controller/ShoppingCartControllerTest.java:90]: incorrect description, this is for user 2 right?

File: src/main/java/mate/project/repository/shoppingcart/ShoppingCartRepository.java [hunk 7]
```diff
@@ -10,7 +11,7 @@
 public interface ShoppingCartRepository extends JpaRepository<ShoppingCart, Long> {
     Optional<ShoppingCart> findByUser(User user);
 
-    Optional<ShoppingCart> findAllByUserId(Long userId);
+    List<ShoppingCart> findAllByUserId(Long userId);
 
     @Query("SELECT sc FROM ShoppingCart sc LEFT JOIN FETCH sc.cartItems WHERE sc.user = :user")
     Optional<ShoppingCart> findCartItemsByUser(@Param("user") User user);
```
Comment (George-Khomiak) [inline src/main/java/mate/project/repository/shoppingcart/ShoppingCartRepository.java:17]: Since you are using custom query you may as well test this method. Create repository tests suit

**Review by George-Khomiak — CHANGES_REQUESTED** [review]
Overall it looks very good, though the test cases are quite basic. In most projects, the majority of tests are unit tests, in comparison to integration tests etc. So, It would be great to include more edge cases, rather than only straightforward functionality tests.

**Comment by jszymanski97** [comment]
Screenshot 2025-02-15 at 19 29 42

**Review by George-Khomiak — APPROVED** [review]
